src/transaction.rs

```diff
@@ -582,6 +582,13 @@ impl<T: Transport> SendFuture<T> {
     }
 }
 
+// NOTE: This is safe since `SendFuture` only requires pinned `SendState` and
```

Comment:
Can you explain how the alternative with trait bounds would work?

Reply:
https://github.com/gnosis/ethcontract-rs/blob/080d884ec904d8574b4c2e7628860daaa5062da0/src/contract/deploy.rs#L72

This is an example. We are basically declaring that only the inner future needs to be `Unpin` for the outer future to be it as well since it’s the only thing that we need pinned references to.

Reply:
And why can't we do that inside SendFuture? How would that leak the SendState type?

Reply:
Because `SendState` is a private type and would appear in a public interface (here the public interface is the trait bounds of a public trait on a public type). Specifically the compiler error is:
```
error[E0446]: private type `transaction::confirm::ConfirmState<T>` in public interface
   --> src/transaction/confirm.rs:145:1
    |
98  |   enum ConfirmState<T: Transport> {
    |   - `transaction::confirm::ConfirmState<T>` declared as private
...
145 | / impl<T> Unpin for ConfirmFuture<T>
146 | | where
147 | |     T: Transport,
148 | |     ConfirmState<T>: Unpin,
149 | | {
150 | | }
    | |_^ can't leak private type
```

Reply:
This would not be necessary if we used pin_project as they would handle the unpinning for us, wouldn't it?

Reply:
It is still a problem with `pin_project` as the generated code also leaks this type. They have a work around to include the implementation in a function which makes the compiler generate a warning instead of an error but then it doesn't play nice with `cargo doc` where types show up that cannot be clicked on (as they have missing documentation).

The work around is just to make `ConfirmState` and `SendState` public, which is what I did. Documentation for these types explicitly state that they are internal types and that they should not be used directly. Also there is no harm in constructing one of these, and we provide no methods to operate on them. Although it isn't 100% ideal, it is an OK solution for now.

Reply:
If you're curious this is what it looks like: https://github.com/taiki-e/pin-project/blob/master/examples/struct-default-expanded.rs#L55-L90